include/mitsuba/core/math.h

```diff
@@ -235,6 +235,45 @@ MI_INLINE Index find_interval(dr::scalar_t<Index> size,
     return dr::binary_search<Index>(1, size - 1, pred) - 1;
 }
 
+/**
+ * \brief This function does the middle search for bisect
+ * 
+ * To mitigate the issue of varying density of floating point numbers on the
+ * number line, the floats are reinterpreted as unsigned integers. As long as
+ * sign of both numbers is the same, this maps the floats to the evenly spaced
+ * set of integers. The middle of these integers ensures that the space of 
+ * numbers is halved on each iteration of the bisection.
+ *
+ * Note that this strategy does not work if the numbers have different sign.
+ * In this case the function always returns 0.0 as the middle.
+ */
+template <typename Scalar>
+Scalar middle(Scalar left, Scalar right) {
+    using ScalarUInt = dr::uint_array_t<Scalar>;
+
+    // if either left or right are infinite, return the sum
+    // which is inf and let the predicate of the
+    // bisection handle it
+    if (!dr::isfinite(left) || !dr::isfinite(right))
+        return left + right;
+
+    // always return zero if left and right have different signs
+    if (dr::sign(left) != dr::sign(right) && left != Scalar(0.0) && right != Scalar(0.0))
+        return Scalar(0.0);
+
+
+    // we reinterpret as UInt using the absolute value, so we store the sign
+    // to reapply after interpreting the result back to Float
+    bool negate = left < Scalar(0.0) || right < Scalar(0.0);
+    left = dr::abs(left);
```

Comment:
I don't understand the need for this. Won't the middle point be the same if these numbers are negative? What happens if you remove the `dr::abs` calls and also remove the ` negate ? -mid : mid` part at the end?

Reply:
I am not sure about the necessity of the absolute value, but since we divide by two, the most significant bit in `mid_int` will always be zero and therefore the sign information is lost and must be reapplied from `negate`.